src/dmd/pragmasem.d

```diff
@@ -0,0 +1,489 @@
+/**
+ * Compiler implementation of the
+ * $(LINK2 http://www.dlang.org, D programming language).
+ *
+ * Copyright:   Copyright (C) 1999-2018 by The D Language Foundation, All Rights Reserved
+ * Authors:     $(LINK2 http://www.digitalmars.com, Walter Bright)
+ * License:     $(LINK2 http://www.boost.org/LICENSE_1_0.txt, Boost License 1.0)
+ * Source:      $(LINK2 https://github.com/dlang/dmd/blob/master/src/dmd/pragmasem.d, _pragmasem.d)
+ * Documentation:  https://dlang.org/phobos/dmd_pragmasem.html
+ * Coverage:    https://codecov.io/gh/dlang/dmd/src/master/src/dmd/pragmasem.d
+ */
+
+module dmd.pragmasem;
+
+import core.stdc.stdio;
+import core.stdc.string;
+
+import dmd.arraytypes;
+import dmd.attrib;
+import dmd.cond;
+import dmd.expression;
+import dmd.expressionsem;
+import dmd.errors;
+import dmd.dinterpret;
+import dmd.dmodule;
+import dmd.dmangle;
+import dmd.dscope;
+import dmd.dsymbol;
+import dmd.dsymbolsem;
+import dmd.globals;
+import dmd.id;
+import dmd.identifier;
+import dmd.mtype;
+import dmd.root.outbuffer;
+import dmd.root.rmem;
+import dmd.statement : Statement, PragmaStatement, ErrorStatement;
+import dmd.statementsem;
+import dmd.tokens : TOK;
+import dmd.dtemplate : getDsymbol;
+import dmd.utf;
+import dmd.utils;
+
+/**
+ * Semanically analyses a PragmaStatement.
+ * Params:
+ *  sc        = scope of evaluation
+ *  ps        = PragmaStatement to analyse
+ */
+Statement pragmaSemantic(Scope* sc, PragmaStatement ps)
+{
+    auto psa = PragmaSemanticAnalysis(sc, ps, null);
+    psa.semantic(ps.ident);
+    return psa.result;
+}
+
+/**
+ * Semanically analyses a PragmaDeclaration.
+ * Params:
+ *  sc        = scope of evaluation
+ *  pd        = PragmaDeclaration to analyse
+ */
+void pragmaSemantic(Scope* sc, PragmaDeclaration pd)
```

Comment:
Note that this function returns a `void`, and the overload of it returns a `Statement`. This certainly looks peculiar, and they shouldn't be overloads if they are doing fundamentally different things.

Reply:
Thats because the statement is needed in the `Statement` semantic, but the declaration is not needed for `Declaration` semantic. Why, I don't know.

Reply:
Because of the distinctly different interfaces, they should have different names, not be overloads.

Overloads are useful for writing generic code, this isn't generic.